apps/hash-graph/lib/type-fetcher/src/fetcher_server.rs

```diff
@@ -0,0 +1,184 @@
+
+#[derive(Clone)]
+pub struct FetchServer;
+
+#[tarpc::server]
+impl Fetcher for FetchServer {
+    async fn fetch_entity_type_exhaustive(
+        self,
+        _context: Context,
+        entity_type_url: String,
+    ) -> Result<TypeFetchResponse, FetcherError> {
+        fetch_entity_type_exhaustive(entity_type_url).await
+    }
+}
+
+#[derive(Debug)]
+struct StreamState {
+    seen: HashSet<String>,
+    queue: VecDeque<String>,
+}
+
+impl StreamState {
+    fn new(seen: HashSet<String>, queue: VecDeque<String>) -> Self {
+        Self { seen, queue }
+    }
+
+    fn with_intitial_state(start: String) -> Self {
+        let mut seen = HashSet::new();
+        seen.insert(start.clone());
+        let mut queue = VecDeque::new();
+        queue.push_back(start);
+
+        Self::new(seen, queue)
+    }
+}
+
+async fn fetch_entity_type_exhaustive(
+    entity_type_url: String,
+) -> Result<TypeFetchResponse, FetcherError> {
+    // let seen: DashSet<String> = DashSet::new();
+
+    let http_client = Client::new();
+    let res = stream::try_unfold(
+        StreamState::with_intitial_state(entity_type_url),
+        |mut state| async {
+            let client = http_client.clone();
+            let next_url = state.queue.pop_front();
+
+            let Some(url) = next_url else { return Ok(None) };
+            let response = fetch_ontology_type(client, url).await?;
+
+            let uris = match response.clone() {
+                FetchedOntologyType::EntityType(schema) => {
+                    let entity_type: EntityType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Entity Type");
+                        FetcherError::TypeParsingError
+                    })?;
+                    traverse_entity_type_references(&entity_type)
+                }
+                FetchedOntologyType::PropertyType(schema) => {
+                    let property_type: PropertyType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Property Type");
+                        FetcherError::TypeParsingError
+                    })?;
+
+                    traverse_property_type_references(&property_type)
+                }
+                FetchedOntologyType::DataType(schema) => {
+                    let data_type: DataType = schema.try_into().map_err(|error| {
+                        tracing::error!(error=?error, "Couldn't convert schema to Data Type");
+                        FetcherError::TypeParsingError
+                    })?;
+
+                    traverse_data_type_references(&data_type)
+                }
+            };
+
+            for uri in uris {
+                // TODO: reconsider how we should handle these "builtin" external types.
+                if uri.starts_with("https://blockprotocol.org/@blockprotocol/types/") {
+                    continue;
+                }
```

Comment:
We should probably do that too, yes. But this specific check is for the builtin BP types such as `Text` and `Number`. Since these are defined in the spec and currently not browsable (for example https://blockprotocol.org/@blockprotocol/types/data-type/boolean/v/1 doesn't lead to anything) I mad this short circuit. I've brought this up with Ciaran